server/src/server.ts

```diff
@@ -104,29 +112,89 @@ function validateTextDocument(textDocument: TextDocument): void {
         end: { line: 1, character: 0 }
       },
       message: err.message,
-      source: "drafter.js"
+      source: parserName
     });
   }
   finally {
     connection.sendDiagnostics({ uri: textDocument.uri, diagnostics });
   }
 }
 
-connection.onDidChangeWatchedFiles((change) => {
-  connection.console.log('We recevied an file change event');
-});
-
 connection.onDocumentSymbol((symbolParam) => {
-  let cat = lodash.head(lodash.filterContent(refractOutput, {element: 'category'}));
-  let cat2 = lodash.filterContent(cat, {element: 'category'});
+  if (currentSettings.parser.exportSourcemap === false) {
+    return Promise.resolve([]); // I cannot let you navigate if I have no source map.
+  }
 
-  let resources = lodash.map(cat2, (ct2) => {return lodash.resources(ct2)});
+  let symbolArray : SymbolInformation[] = [] ;
+
+  const textDocument = documents.get(symbolParam.textDocument.uri);
+  const documentLines = textDocument.getText().split(/\r?\n/g);
+
+  let mainCategory = lodash.head(lodash.filterContent(refractOutput, {element: 'category'}));
+
+  // The first category should always have at least a title.
+  const title = lodash.get(mainCategory, 'meta.title');
+  if (typeof(title) !== 'undefined') {
+    const lineReference = refractUtils.createLineReferenceFromSourceMap(title.attributes.sourceMap, textDocument.getText(), documentLines);
+    symbolArray.push(SymbolInformation.create(
+      title.content,
+      SymbolKind.Package,
+      Range.create(
+        lineReference.startRow,
+        lineReference.startIndex,
+        lineReference.endRow,
+        lineReference.endIndex
+      )
+    ));
+  }
 
-  const symbolArray = lodash.map(lodash.flatten(resources), (resource) => {
-    return SymbolInformation.create(resource.meta.title, SymbolKind.Property, Range.create(1,1,1,1), "", "");
-  });
 
-  return Promise.resolve<SymbolInformation[]>(symbolArray);
+  [{
+    query: {
+    "element": "category",
+      "meta": {
+        "classes": [
+          "resourceGroup",
+        ],
+      },
+    },
+    symbolType: SymbolKind.Namespace
+  }, {
+      query: {"element": "resource"},
+      symbolType: SymbolKind.Method
+  }
+  ].forEach(({query, symbolType}) => {
+    const queryResults = refractUtils.query(refractOutput, query);
+
+    symbolArray.push(...lodash.map(queryResults, (queryResult) => {
+      let sourceMap = lodash.get(queryResult, 'attributes.sourceMap',
+        lodash.get(queryResult, 'meta.title.attributes.sourceMap',
+          lodash.get(queryResult, 'attributes.href.attributes.sourceMap',
+            lodash.get(queryResult, 'content[0].attributes.method.attributes.sourceMap')

```

Comment:
Just a note, in looking at this, I think that all four of these `get` functions will be executed each time.